HW0/P5/P5.txt

```diff
@@ -1 +1,21 @@
 # Your answers here.
+
+1. 255 sec. Since we have 256 bags, adding two numbers cost 1 second, so the total will be 255 seconds. 
+
+2. Yes, they can verify the total sum faster than the original cashier by herself. They can do this in 34 seconds. 
+
+First, we can split 256 bags into 8 groups, each group has 32 bags. For one person, she can count 32 bags in 31 seconds. Now we have 8 bigger bags now. Then split the 8 bigger bags into 4 groups, each group has 2 bags, so 4 people can count in 1 second. Now we have 4 bigger bags that use 1 second to count as well. Finally, we have two bags that use 1 second as well. Therefore, total time is 31 + 1 + 1 + 1 = 34 sec.
+
+3. 8 seconds.
+
+Same idea but now we have infinity people to count. From 256 -> 128 -> 64 -> 32 -> 16 -> 8 -> 4 -> 2 which is 8 seconds.
+
+4. Please check P5.py and P5.png
+
+5. 15 seconds. Let’s say a complete "action” includes sum step and communicate step. Since all people can communicate and sum at the same time and each step only takes 1 sec, so the fastest method is the one that requires the least actions. Thus, we should get the as many people involved in each communicate or sum step as possible to reduce actions needed. We split 256 bags into 128 groups, each group has 2 bags. The reason we split like that because communication time is 1 second between 2 people, the fastest method is to let the most people do this work, which is making the counting time lowest. From 256—>128, the counting time is 1 second, the commutation time is 1 second and so on. However, from 2->1, we have counting time, but do not have commutation time. Therefore, the time is 2 * 7 + 1 = 15 seconds. 
+
+6. 1 bag: For total 256 bags, we ignore the handing time first, each person takes 1 bag, so we have 256 groups. Same idea as problem 5, but a little bit different is 256 people should communicate first, it takes an extra 1 second. So the time without handing time is 16 seconds. Now for the handing time, from 256—>1, it takes 8 seconds, but we need 1 extra second to pass bags to cashier. Therefore, the total time will be 16 + 8 + 1 = 25 sec
```

Comment:
My understanding of this problem was that you had to hand each employee one bag, and you can only hand them bags one at a time. It also seems that once the bags are handed out from you to the employees, no more handing needs to be done. Thus, when you distribute the bags serially - one bag to one worker - it should take 256 seconds to distribute.